remote.c

```diff
@@ -2163,6 +2163,150 @@ do_del_tsig(RES* ssl, xfrd_state_type* xfrd, char* arg) {
+		(void)ssl_printf(ssl, "error: can not drop the currently active cookie secret\n");
+		return;
+	}
+	if(!nsd->options->cookie_secret_file || !nsd->options->cookie_secret_file[0]) {
+		(void)ssl_printf(ssl, "error: no cookie secret file configured\n");
+		return;
+	}
+	if(!cookie_secret_file_dump(ssl, nsd)) {
+		(void)ssl_printf(ssl, "error: writing to cookie secret file: \"%s\"\n",
+				nsd->options->cookie_secret_file);
+		return;
+	}
+	drop_cookie_secret(nsd);
+	(void)cookie_secret_file_dump(ssl, nsd);
+	task_new_drop_cookie_secret(xfrd->nsd->task[xfrd->nsd->mytask],
+	    xfrd->last_task);
+	xfrd_set_reload_now(xfrd);
+	send_ok(ssl);
+}
+
+static void
+do_add_cookie_secret(RES* ssl, xfrd_state_type* xrfd, char* arg) {
+	nsd_type* nsd = xrfd->nsd;
+	uint8_t secret[NSD_COOKIE_SECRET_SIZE];
+
+	if(*arg == '\0') {
+		(void)ssl_printf(ssl, "error: missing argument (cookie_secret)\n");
+		return;
+	}
+	if(strlen(arg) != 32) {
+		(void)ssl_printf(ssl, "invalid cookie secret: invalid argument length\n");
+		(void)ssl_printf(ssl, "please provide a 128bit hex encoded secret\n");
+		memset(arg, 0xdd, strlen(arg));
+		return;
+	}
+	if(hex_pton(arg, secret, NSD_COOKIE_SECRET_SIZE) != NSD_COOKIE_SECRET_SIZE ) {
+		(void)ssl_printf(ssl, "invalid cookie secret: parse error\n");
+		(void)ssl_printf(ssl, "please provide a 128bit hex encoded secret\n");
+		memset(arg, 0xdd, strlen(arg));
+		return;
+	}
+	if(!nsd->options->cookie_secret_file || !nsd->options->cookie_secret_file[0]) {
+		(void)ssl_printf(ssl, "error: no cookie secret file configured\n");
+		memset(arg, 0xdd, strlen(arg));
+		return;
+	}
+	if(!cookie_secret_file_dump(ssl, nsd)) {
+		(void)ssl_printf(ssl, "error: writing to cookie secret file: \"%s\"\n",
+				nsd->options->cookie_secret_file);
+		memset(arg, 0xdd, strlen(arg));
+		return;
+	}
+	add_cookie_secret(nsd, secret);
+	(void)cookie_secret_file_dump(ssl, nsd);
+	task_new_add_cookie_secret(xfrd->nsd->task[xfrd->nsd->mytask],
+	    xfrd->last_task, arg);
+	memset(arg, 0xdd, strlen(arg));
+	xfrd_set_reload_now(xfrd);
+	send_ok(ssl);
+}
+
+static void
+do_print_cookie_secrets(RES* ssl, xfrd_state_type* xrfd, char* arg) {
+	nsd_type* nsd = xrfd->nsd;
+	char secret_hex[NSD_COOKIE_SECRET_SIZE * 2 + 1];
+	int i;
+	(void)arg;
+
+	/* (void)ssl_printf(ssl, "cookie_secret_count=%zu\n", nsd->cookie_count); */
+	for(i = 0; (size_t)i < nsd->cookie_count; i++) {
+		struct cookie_secret const* cs = &nsd->cookie_secrets[i];
+		ssize_t const len = hex_ntop(cs->cookie_secret, NSD_COOKIE_SECRET_SIZE,
+		                             secret_hex, sizeof(secret_hex));
+		(void)len; /* silence unused variable warning with -DNDEBUG */
+		assert( len == NSD_COOKIE_SECRET_SIZE * 2 );
+		secret_hex[NSD_COOKIE_SECRET_SIZE * 2] = '\0';
+		if (i == 0)
+			(void)ssl_printf(ssl, "active : %s\n",  secret_hex);
+		else if (nsd->cookie_count == 2)
+			(void)ssl_printf(ssl, "staging: %s\n",  secret_hex);
+		else
+			(void)ssl_printf(ssl, "staging[%d]: %s\n", i, secret_hex);
+	}
```

Comment:
If you want to clear the array, here `secret_hex` is the array to clear.

Reply:
resolved in aa3bca3